### ISSUE TITLE
rom-rails#success? does not work with ROM::ConstraintError

### ISSUE THREAD

**Issue by kwando** [rom-rb/rom#204]
```ruby
irb(main):013:0> ApprovalForm.build(description: 'hello world', incident_id: 106694).save.success?
  ROM[mysql] (1.5ms)  INSERT INTO `approvals` (`description`, `incident_id`) VALUES ('hello world', 106694)  
NoMethodError: undefined method `any?' for #<ROM::SQL::ConstraintError:0x007f95e4e59468>
        from /Users/kwando/projects/work/sms_api2/vendor/bundle/ruby/2.1.0/gems/rom-rails-0.3.1/lib/rom/rails/model/form.rb:116:in `success?'
```

**Comment by cflipse** [comment]
Test to reproduce is here:
https://github.com/cflipse/rom-rails/commit/a17012030c046634fb34661ac68c5c1bb3b89a2f

Looking at this, I'm not actually sure that this is a rom-rails error.  I think the "actual" bug is that the exception is leaking out of the `try` block, rather than the expected `ROM::Commands::Result::Failure` object

**Comment by cflipse** [comment]
So, I can make that specific error go away with a small code change, _but_

Here's the thing:  ActiveRecord lets validation errors bubble all the way up, typically.  You violate a database constraint, it explodes immediately.  This is kind-of good, as you get an error close to the failure.  ROM swallows the constraint errors and return them in a failed command data object, and you check / verify a few lines down.

Unfortunately, that error accessor is _also_ the place where we put validation failures.  Same attribute, different APIs.  We can catch the constraint failure, and successfully report that the command has failed ... and rails will, in most cases, re-render the form, and attempt to display the validation error messages.  Which will explode with another no method error, this time for `[]`.  And it'll be even further away from the source.

It kind of feels to me like the issue is that we conflate validation failures and persistence failures in the same spot, and the solution lies closer to `CommandRegistry#try`  .. I'm not entirely sure.  @solnic, thoughts?

**Comment by solnic** [comment]
This is a super important subject to discuss. My way of thinking about it initially was that we're gonna treat specific constraint violations as validation errors too that were raised by the db backend (sequel in our case). Then you are free to handle them however you want, ie you may re-raise the error if you want.

I'm not sure what the problem is with re-rendering the form. What `no method` error for `[]`? rom-rails should wrap those errors with whatever is needed so that they can be handled easily. I wouldn't go as far as trying to treat them as the same kind of a validation error form the ruby side (ie unique constraint validation vs ruby AM uniqueness validation), because semantics are different. But I think it should be possible to just handle constraint errors on the UI layer by simply showing an error message that it failed and please try again. We're talking about edge cases here but they happen, right.

Your code can handle that error, you've got original error and you've got command error which wraps it. So it's a matter of handling it properly.

**Comment by cflipse** [comment]
in progress:  https://github.com/cflipse/rom-rails/tree/constraint-errors

If you pull that branch and run the specs, you'll get:  
```
  1) Form #errors recovers from database errors
     Failure/Error: expect(result.errors[:email]).to eq []
     NoMethodError:
       undefined method `[]' for #<ROM::SQL::ConstraintError: ROM::SQL::ConstraintError>
```

So, on a failure, the form assigns and treats the error as the value coming back from command.try, which nests whatever real failure it got.  Sometimes, that's a validation error.  Sometimes, that's the database constraint.

The form is expecting the nested command error to conform to AM::Validations, which the constraint does not do, nor should it.  For those violations, I think all we've really got is a string message, and we can't assign it to the field or fields that triggered the violation.

I suspect what we're looking at is more along the lines of having the forms using an error proxy, that can look at both "types" of failures,  and present them in a non-explody kind of way.  That might also clean up the 'pre-validation' error blocks that I had to do to make the forms render cleanly in the first place.

**Comment by solnic** [comment]
Yes that was my idea, not to choke on those errors and handle them gracefully and simply treat db constraints as your second defense against corrupted data.

**Comment by cflipse** [comment]
fixed in rom-rb/rom-rails/pull/35
